docs/source/howto/workflows.rst

```diff
@@ -4,14 +4,217 @@
 How to run multi-step workflows
 *******************************
 
-
 .. _how-to:workflows:write:
 
 Writing workflows
 =================
 
-`#3991`_
+A workflow in AiiDA is a :ref:`process <topics:processes:concepts>` that calls other workflows and calculations and optionally *returns* data and as such can encode the logic of a typical scientific workflow.
+Currently, there are two ways of implementing a workflow process:
+
+ * :ref:`work functions<topics:workflows:concepts:workfunctions>`
+ * :ref:`work chains<topics:workflows:concepts:workchains>`
+
+Here we present a brief introduction on how to write both workflow types.
+
+.. note::
+
+    For more details on the concept of a workflow, and the different between a work function and a work chain, please see the corresponding :ref:`topics section<topics:workflows:concepts>`.
+
+Work function
+-------------
+
+A *work function* is a process function that calls one or more calculation functions and *returns* data that has been *created* by the calculation functions it has called.
+Writing a work function that can be stored in the provenance simply involves writing a Python function that calls a set of calculation functions with the desired logic and decorating it with the ``@workfunction`` decorator:
+
+.. code-block:: python
+
+    from aiida.engine import calcfunction, workfunction
+    from aiida.orm import Int
+
+    @calcfunction
+    def add(x, y):
+        return x + y
+
+    @calcfunction
+    def multiply(x, y):
+        return x * y
+
+    @workfunction
+    def add_and_multiply(x, y, z):
+        sum = add(x, y)
+        product = multiply(sum, z)
+        return product
+
+    result = add_and_multiply(Int(1), Int(2), Int(3))
+
+It is important to reiterate here that the ``@workfunction``-decorated ``add_and_multiply`` function does not *create* any new data nodes.
+The ``add`` and ``multiply`` calculation functions create the ``Int`` data nodes, all the work function does is *return* the results of the ``multiply`` calculation function.
+Moreover, both calculation and workflow functions can only accept and return data nodes, i.e. instances of classes that subclass the ``Data`` class.
+
+Work chain
+----------
+
+When the workflow you want to run that is more expensive and complex, it is better to write a *work chain*.
+Writing a work chain in AiiDA requires creating a class that inherits from the ``WorkChain`` class and defines the work chain.
+Below is an example of a work chain that takes three integers as inputs, multiplies the first two and then adds the third to obtain the final result:
+
+.. code-block:: python
+
+    from aiida.orm import Code, Int
+    from aiida.engine import calcfunction, WorkChain, ToContext
+    from aiida.plugins.factories import CalculationFactory
```

Comment:
*SPH hyper-nitpick-mode-engaged*: I try to keep imports alphabetically sorted. If they get larger, makes it easy to quickly scan it

Reply:
I can see the value in that ^^

Reply:
However, I'll keep the example as is for now, since we'll probably include the code directly once this is merged into `develop`?

Reply:
Fine to keep here for now. We will take care of it in #4128 once `docs-revamp` is merged in `develop`